src/lattices/abstract.jl

```diff
@@ -0,0 +1,66 @@
+"""
+Abstract definition of a lattice.
+"""
+abstract type AbstractLattice end
+
+# AbstractLattice interface: mandatory
+# TODO: This needs to be updated. There must be a general way to access sites and bonds.
+"""
+    length(l::AbstractLattice)
+
+Number of lattice sites.
+"""
+@inline Base.length(l::AbstractLattice) = error("Lattice $(typeof(l)) doesn't implement `length`.")
+
+
+
+
+
+
+# AbstractLattice interface: optional
+"""
+    neighbors(l::AbstractLattice, drop_repeats::Val{true})
+    neighbors(l::AbstractLattice, drop_repeats::Val{false})
+
+Returns an iterator over bonds, given as tuples (source index, target index). If
+`drop_repeats = Val(false)` bonds are assumed to be directed, i.e. both
+`(1, 2)` and `(2, 1)` are included. If `drop_repeats = Val(true)` bonds are
+assumed to be undirected, i.e. `(1, 2)` and `(2, 1)` are assumed to be
+equivalent and only one of them will be counted.
+"""
+@inline function neighbors(l::AbstractLattice, drop_repeats::Val{false})
+    (
+        [source, target] for source in 1:length(l)
```

Comment:
Why return a vector (not a tuple)?

Reply:
Because the other two methods return vectors. I thought it'd be better to be consistent. (Yet two return iterators of vectors and one a vector integers...)